### PR TITLE
Allow docker.yml action to run within developer fork.

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

_Select the type of your PR_

- Bugfix

### Description

The docker.yml github hub flow action should only run when properly configured.   Makes its execution conditional and document what a developer would need to do to use the action in their own fork.

This change moves to using repository variables, so the workflows execution can be made conditional on it.
It also adopts container registry neutral terminology, to help users using docker.io or ghcr.io etc.

Why: The github action was failing when run in my fork ending with an error message.  This created distracting emails.

```
buildx failed with: ERROR: failed to solve: failed to push quay.io/kroxylicious/kroxylicious-developer:0.3.0-SNAPSHOT: unexpected status from POST request to https://quay.io/v2/kroxylicious/kroxylicious-developer/blobs/uploads/: 401 UNAUTHORIZED
```

### Additional Context

_Why are you making this pull request?_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

**Review by SamBarker — APPROVED** [review]
LGTM, assuming one has QUAY_USERNAME defined in the fork. 

Worth adding that to the dev_guide?

**Comment by k-wall** [comment]
> LGTM, assuming one has QUAY_USERNAME defined in the fork.
> 
> Worth adding that to the dev_guide?

Yes, I think that's useful.

I should make a further change to the action to follow this pattern, so that the action only runs if the required secrets are present.  This would mean that the action doesn't break fail on the developer's fork.    WDYT?

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-using-secrets

File: .github/workflows/docker.yml
Comment (k-wall) [inline .github/workflows/docker.yml:14]: Note that Github doesn't support conditions using secret values at all.  At the `jobs.<job_id>.if` level, it is legal to refer to `variables.`.  That was my motivation for moving the   QUAY_USERNAME from secret to variable.

https://docs.github.com/en/actions/learn-github-actions/contexts#context-availability

File: .github/workflows/docker.yml
Comment (k-wall) [inline .github/workflows/docker.yml:44]: I needed to factor out QUAY_ORG too as QUAY_USERNAME isn't necessarily the same value (and won't be if using a robot account).

**Comment by k-wall** [comment]
If already updated kroxylicious to have a variable QUAY_USER.  Once this PR is merged, I'll remove the secret as it will then be redundant.

File: DEV_GUIDE.md
Comment (SamBarker) [inline DEV_GUIDE.md:324]: nit: I wonder if thats too strong a statement. 

```suggestion
All CI [workflows](.github/workflows) defined by the project are expected to execute within the context of a fork, apart from [docker workflow](.github/workflows/docker.yml).
```

File: DEV_GUIDE.md
Comment (SamBarker) [inline DEV_GUIDE.md:328]: Is it worth giving an example or outlining what this represents?

e.g. 
```suggestion
* `QUAY_ORG` variable - your Quay organisation. Will be used to construct the addresses used for pushing and pulling images.
```

File: DEV_GUIDE.md
Comment (SamBarker) [inline DEV_GUIDE.md:329]: We should probably default to assuming robot accounts as a hint to better security practice by default.

File: DEV_GUIDE.md
Comment (SamBarker) [reply DEV_GUIDE.md:328]: Oh I see you do that further down. Feel free to ignore...

**Review by SamBarker — APPROVED** [review]
LGTM a nice change to improve our developer experience.

As usual feel free to ignore the nits.

File: DEV_GUIDE.md
Comment (SamBarker) [inline DEV_GUIDE.md:332]: Not strictly part of this PR but should we also include some hints about registering for a free Quay account? We probably can't get away with assuming everyone already has one...

File: DEV_GUIDE.md
Comment (robobario) [inline DEV_GUIDE.md:326]:
```suggestion
To enable the docker workflow, you need to configure two repository [variables](https://docs.github.com/en/actions/learn-github-actions/variables)
and one repository [secret](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions).
```

File: DEV_GUIDE.md
Comment (k-wall) [reply DEV_GUIDE.md:332]: I can add that near the prerequisites section. 

Probably best not to assume Quay.io. User could use docker.io or ghcr or something else.

With that in mind, I think to rejig the variables/secret to use container registry neutral language

* `QUAY_USER` -> `REGISTRY_USERNAME` 
* `QUAY_TOKEN` -> `REGISTRY_TOKEN`
* `QUAY_ORG` -> `REGISTRY_DESTINATION` (quay.io/kroxylicious/kroxylicious-developer) (
* REGISTRY_SERVER eg. docker.io or quay.io

justification for names:

1. both `docker login` and `podman login` CLIs use the term `username` rather than `user`
2. podman push CLI  term uses term `destination`   (docker push uses `name` but that too general)
3. podman login uses term `server` docker login uses the term `registry`.  

How's that sound?

File: DEV_GUIDE.md
Comment (k-wall) [reply DEV_GUIDE.md:329]: docker.io doesn't have the same robot account concept as quay.io.  Instead it lets you generate access token(s) that are used in place of your password together with your regular username.

https://hub.docker.com/settings/security

File: DEV_GUIDE.md
Comment (SamBarker) [reply DEV_GUIDE.md:332]: Yeah that sounds very reasonable

Comment (k-wall) [reply DEV_GUIDE.md]: @SamBarker I made that change.   could you re-review please  I'd like to refactor the container registry interaction in `/scripts/run-with-strimzi.sh`  etc to follow the same pattern, but this is better in a separate PR.

**Review by SamBarker — APPROVED** [review]

**Pull request merged**
